etcd v3 initial change

hongchaodeng opened the pull request:
This serves as a follow up on #22448.

The conversation is good! Let's keep it up while working out some real code.

I'm initiating a few things here:
- bump up etcd godep. The etcd would support both V2 and V3 API for us to do the work. Minor fix is done on EtcdtestServer.
- I set up the structure of etcd3 helper and implements Set(), Get().

hongchaodeng committed:
etcd V3 helper: implements Set(), Get()

wojtek-t commented:
@hongchaodeng - can you please move Godeps update to a separate PR? It would be so much easier to review this PR then, because github doesn't support well PRs with so many files...
 
Also, I assume that with upgraded Godeps, etcd v2 in current setup will still work, right?

timothysc commented:
@hongchaodeng could you squash the helper commits please, I can review 2-day.

wojtek-t commented:
@timothysc - I didn't review this, just because we don't want to have it in 1.2. We are explicitly not merging big changes that are not going to be picked to 1.2 now.

timothysc commented:
@wojtek-t no worries, I had some time so just getting back into the fray, and wanted to keep an eye on this one.

hongchaodeng commented:
I am constantly getting the error:
```
output binaries contain bad symbols "testing[.]"
!!! Error in ./hack/../hack/verify-symbols.sh:28
  '"${KUBE_ROOT}/hack/after-build/verify-symbols.sh"' exited with status 1
Call stack:
  1: ./hack/../hack/verify-symbols.sh:28 main(...)
Exiting with status 1
FAILED   ./hack/../hack/verify-symbols.sh
```

@timothysc Can you help with that?

hongchaodeng commented:
Quick update:

- Fixed build problems
- Addressed comments 
- re. separating out godeps, it's not possible. AFAIK, "verify-godeps" is gonna fail because some deps are not being used without the the v3 helper implementation

/cc @wojtek-t @timothysc

hongchaodeng commented:
The failure happened in metrics grabber tests.
It looked like:
```
MetricsGrabber
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/metrics_grabber_test.go:158
  should grab all metrics from API server. [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/metrics_grabber_test.go:100

  Expected
      <sets.String | len:39>: {
          "etcd_storage_events_total": {},
          "etcd_storage_db_compaction_pause_duration_milliseconds_sum": {},
          "etcd_storage_pending_events_total": {},
          "etcd_snapshot_save_marshalling_durations_seconds_sum": {},
          "etcd_storage_index_compaction_pause_duration_milliseconds_bucket": {},
          ....
      }
  to be empty
```

Those are etcd server metrics. Any idea why? @gmarek

magicwang-cn commented on pkg/storage/etcd/v3_helper.go:
```diff
@@ -0,0 +1,216 @@
+	key = prefixEtcdKey(key, h.pathPrefix)
+
+	// Note: if rev = 0, it is equivalent to create.
+	txnResp, err := h.client.KV.Txn(ctx).If(
```
the client should be closed after using it? like [etcdv3api](https://github.com/coreos/etcd/tree/master/clientv3)

xiang90 replied:
i think this client is a long running one. you only need to close the client if you do not plan to use it anymore.

magicwang-cn replied:
ok. in the high concurrency scenarios, will multi-clients work well than single-long-running client?

xiang90 replied:
it does not really matter if you have multiple clients or one. But you probably want to enable concurrent requests.

hongchaodeng commented:
The travis-ci seems to have problems with
```
FAILED   hack/../hack/verify-godep-licenses.sh
```

But I couldn't get anything new by running `./hack/update-godep-licenses.sh`, nor running into error in `./hack/verify-godep-licenses.sh` locally.

What's the issue? Anyone has any idea?

ixdy commented:
@david-mcmahon @karlkfi @eparis any idea what's happening with the godep licenses check here? I don't see why it's intermittently passing/failing on Jenkins or Travis.

david-mcmahon commented:
Is there a repro case we can identify/investigate?  How does one simulate the environment where this is happening?

ixdy commented:
On Jenkins we're using the `kubekins-test` image, so it should be fairly reproducible (see https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/gotest-dockerized.sh#L36-L46).

I have no idea why this failed and then passed on Travis.

eparis commented:
I can't find the string 'This package and the code it generates requires at least Go 1.4.' in the kube repo anywhere accept this LICENSE file (why and to whom is this monstrosity useful?)

So I'm truly clueless what this is at the moment.

eparis commented:
That text (which absolutely doesn't belong) apparently lives in:
https://github.com/golang/protobuf/blob/master/README.md

david-mcmahon commented:
That line comes from https://github.com/gogo/protobuf/blob/master/README.
It and everything else update-godep-licenses.sh collects has been requested by and is useful to lawyers.

david-mcmahon commented:
Whether it belongs or not though is not the immediate issue - which is why is jenkins flaking on Jenkins/travis.

eparis commented:
that's fair (and I guess it does belong, the README.md uses the word 'Copyright')

david-mcmahon commented:
Yep, I wrote update-godep-licenses.sh to satisfy the lawyers and to not become a maintenance burden on us.  It paints with a fairly wide brush and collects some bits we may not need, but it shouldn't require any handholding or hardcoding.

eparis commented:
Would be a whole lot easier to concatenate the Godeps/_workspace/src/*/LICENSE(ish) files, and fix godeps if it is missing including things we need....   Again, off topic.

david-mcmahon commented:
I went that route.  See the (not used) --create-missing flag.  We can't use that because _workspace is auto-generated and we can add things there that godep doesn't put there itself. 
During the course of creating this tool, we thought about creating a parallel tree to store files godep doesn't get itself, but we left that as an option for later.  If you want to do that, go for it.

eparis commented:
CI falls down on verify-godeps.sh if you add something (or remove something) that should be auto-generated. But alas, not everything you are collecting is present (and things can get merged which break CI).

So the real question is how can it 'occasionally' fail to notice the readme?

eparis commented:
Any github failure other than the expected 404 would result in the symptom, right?

eparis commented:
Since its obviously not reproducible makes me think of a transient raw.githubusercontent.com error, which we might be able to minimize problems just adding `--retry 10` or something like that?  Wish there were an easy way to distinguish 404 from other errors  :-(.  But at least 404 doesn't trigger curl's retry mechanism, so it should be safe...

david-mcmahon commented:
Perhaps github rate limiting affecting mostly/only CI/Travis is the cause.

david-mcmahon commented:
--retry a good idea.  Will send you a PR.

eparis commented:
Even more horrible monstrosity:
```
curl --fail -sS --retry 5 https://${package_root_url}/master/${f})" 2>&1 | grep -v 'curl: (22) The requested URL returned error: 404 Not Found'
```

So if we get a transient failure it will actually show up in the output file...

hongchaodeng commented:
After a long fight, all tests are green now.

@wojtek-t Can you help review and give feedback on this?

timothysc commented on pkg/storage/etcd/v3_helper.go:
```diff
@@ -0,0 +1,216 @@
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
```
All of these functions still need implementations.

wojtek-t replied:
Yes - definitely.

timothysc commented on pkg/storage/etcd/v3_helper.go:
```diff
@@ -0,0 +1,216 @@
+	panic("unimplemented")
+}
+
+func prefixEtcdKey(key, prefix string) string {
```
I believe there is a generic function for this already.

xiang90 replied:
also i think this is not etcd specific. probably just make this keyWithPrefix.

hongchaodeng replied:
done.

timothysc commented on pkg/storage/etcd/v3_helper.go:
```diff
@@ -0,0 +1,216 @@
+	return path.Join(prefix, key)
+}
+
+func extractRevision(versioner storage.Versioner, obj runtime.Object) (int64, error) {
```
Same comment about pushing this out.

wojtek-t replied:
I'm not aware of anything, to be honest.

hongchaodeng replied:
I think Tim was talking in this context:
https://github.com/hongchaodeng/kubernetes/commit/21f089f27f94440ce6713630168f368c411cfaa1#commitcomment-16690709

wojtek-t commented on pkg/storage/etcd/v3_helper.go:
```diff
@@ -34,6 +33,7 @@ import (
 
 var (
 	ErrNotFound = errors.New("etcd: key not found")
+	ErrTestFail = errors.New("etcd: test failed")
```
Can you please change it to:
"etcd: compare failed"
?

"test failed" is suggesting that it is test-related, which it isn't

hongchaodeng replied:
Let me reply to all your comments w.r.t. these errors.

What's the problem?
Those are etcd errors and would be handled specially. See [here](https://github.com/kubernetes/kubernetes/blob/c505a5d49d35011abcb9c4fa853eb33b7b1efb6e/pkg/storage/etcd/util/etcd_util.go#L29-L65).
However, etcd3 wouldn't have return those errors.

How to resolve them?
The simplest way is to make etcd2 errors and return them. That's my plan. What's your opinion?

/cc @xiang90

hongchaodeng replied:
Sorry. Let me rectify my previous answer.

The correct solution should be we should have storage specific errors. Storage would be hiding etcd2 or etcd3 specific errors.

wojtek-t commented on pkg/storage/etcd/v3_helper.go:
```diff
@@ -0,0 +1,216 @@
+)
+
+var (
+	ErrNotFound = errors.New("etcd: key not found")
```
I'm slightly against having such errors - "key not found" without mentioning the "key" is not very useful.

wojtek-t commented:
I've added few comments.

@hongchaodeng Can't we merge first two commits as a separate PR? We could do it now.

hongchaodeng commented:
@wojtek-t 
Separate out the first two commits: https://github.com/kubernetes/kubernetes/pull/23194

Some deps are missing there because we only use them in v3 helper commits. But they are pretty minor.

hongchaodeng committed:
add custom storage error

hongchaodeng commented:
Add last commit "add custom storage error".
It abstracts away etcd-specific implementation of errors.
It's up to implementation re. how to parse error details and return storage errors.

/cc @xiang90 @wojtek-t @timothysc

xiang90 commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -711,3 +710,21 @@ func (h *etcdHelper) addToCache(index uint64, obj runtime.Object) {
 	}
 }
+
+func convertToStorageErrorIfPossible(err error, key string, rv uint64) error {
```
just ToStorageErr?

xiang90 replied:
is this an etcd specific func? or we might add other potential kv error types into switch cases?

hongchaodeng replied:
"storage/" handles only four errors in special. See [here](https://github.com/kubernetes/kubernetes/blob/a428246960b83eb568bef1056dcc30d8ca0254bd/pkg/storage/errors.go#L23-L45).

Not sure about other error types ATM.

hongchaodeng committed:
godep: add clientv3 and integration

hongchaodeng commented:
Rebased and addressed comments.

hongchaodeng commented:
I moved out the storage error commit into another PR (https://github.com/kubernetes/kubernetes/pull/23295) for easy review. PTAL.

hongchaodeng commented:
This PR is way too long. 

It's github's problem -- They should have better ways to show long discussion.

Anyway, I'm gonna closing this. Once https://github.com/kubernetes/kubernetes/pull/23295 is merged, I will rebase and submit another PR with more KV methods and improvement of code quality.

hongchaodeng commented:
Created a new PR https://github.com/kubernetes/kubernetes/pull/23387
